src/utils/generator.js

```diff
@@ -253,10 +253,17 @@ class ModifiedBindingEntry extends GeneratorEntry {
       containingGenerator === this.containingGenerator,
       "This entry requires effects to be applied and may not be moved"
     );
-    invariant(this.modifiedBinding.value === this.newValue);
+    invariant(
+      this.modifiedBinding.value === this.newValue,
+      "ModifiedBinding's value has been changed since last visit."
+    );
     let [residualBinding, newValue] = context.visitModifiedBinding(this.modifiedBinding);
-    invariant(this.residualFunctionBinding === undefined || this.residualFunctionBinding === residualBinding);
+    invariant(
+      this.residualFunctionBinding === undefined || this.residualFunctionBinding === residualBinding,
+      "ResidualFunctionBinding has been changed since last visit."
+    );
     this.residualFunctionBinding = residualBinding;
+    this.modifiedBinding.value = newValue;
```

Comment:
Is this change really needed? Looks a bit scary to me. I'd rather do less state mutation on the side than more.

Reply:
Without it, we run into the follow problem:

In the provided example
```
(function () {
  let a = __abstract("number", "a");
  let b, c;
  global.f = function() {
    b = a + 42;
    c = a + 42;
  }
  __optimize(f);
})();
```
when the `ModifiedBindingEntry` does the `.visitModifiedBinding`, there is a `.visitEquivalentValue(newValue)` call on the new value, and the returned newValue may not match the original (it's replaced by the "equivalent" value if one is found). This happens to us in this test case because of the 
```
b = a + 42;
c = a + 42;
```
since the `a + 42` look the same. The assignment realigns the state to what is expected by the invariant assertion. 

This problem actually seems to be at least somewhat independent of the problem caused by the cache/ResidualFunctionBinding, but it was in the same code area and also causes the code to crash. 

It seems that the more fundamental issue is that we're re-traversing a the heap, which has some sort of assumption that it won't be re-traversed.

Reply:
> which has some sort of assumption that it won't be re-traversed.

Yes, that's the underlying problem here...

While the introduced assignment makes the invariant violation go away, it changes what happens during visiting, which is bad. 

The right thing to do would be to stop mutating state on the side in these GeneratorEntry's, they should be just values. Instead, build up new data structure in the visitor, a mapping of GeneratorEntry to something --- whatever mutable state there was, and pass that on to the next phase that really needs it, and just throw it away after the `ResidualHeapGraphVisitor`.

Reply:
@yinghuitan was saying then that it might be easier to detect and reject calls to `__optimize()` when the `--heapGraphFilePath` flag is given, since the two haven't been made to work together yet. Would that be a more relevant solution?